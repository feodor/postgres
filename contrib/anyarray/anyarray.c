/*-------------------------------------------------------------------------
 * 
 * pg_prewarm.c
 *		prewarming utilities
 *
 * Copyright (c) 2010-2014, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		contrib/anyarray/anyarray.c
 *
 *-------------------------------------------------------------------------
 */

#include "anyarray.h"

#include <access/hash.h>
#include <access/htup_details.h>
#include <access/nbtree.h>
#include <catalog/pg_am.h>
#include <catalog/pg_cast.h>
#include <catalog/pg_collation.h>
#include <commands/defrem.h>
#include <utils/catcache.h>
#include <utils/lsyscache.h>
#include <utils/syscache.h>

PG_MODULE_MAGIC;

static Oid
getAMProc(Oid amOid, Oid typid)
{
	Oid		opclassOid = GetDefaultOpClass(typid, amOid);

	Assert(amOid == BTREE_AM_OID || amOid == HASH_AM_OID);

	if (!OidIsValid(opclassOid))
	{
		typid = getBaseType(typid);
		opclassOid = GetDefaultOpClass(typid, amOid);

		if (!OidIsValid(opclassOid))
		{
			CatCList	*catlist;
			int			i;

			/*
			 * Search binary-coercible type
			 */
			catlist = SearchSysCacheList(CASTSOURCETARGET, 1,
										 ObjectIdGetDatum(typid),
										 0, 0, 0);
			for (i = 0; i < catlist->n_members; i++)
			{
				HeapTuple		tuple = &catlist->members[i]->tuple;
				Form_pg_cast	castForm = (Form_pg_cast)GETSTRUCT(tuple);

				if (castForm->castfunc == InvalidOid && castForm->castcontext == COERCION_CODE_IMPLICIT)
				{
					typid = castForm->casttarget;
					opclassOid = GetDefaultOpClass(typid, amOid);
					if(OidIsValid(opclassOid))
						break;
				}
			}

			ReleaseSysCacheList(catlist);
		}
	}

	if (!OidIsValid(opclassOid))
		return InvalidOid;

	return get_opfamily_proc(get_opclass_family(opclassOid),
							 typid, typid,
							 (amOid == BTREE_AM_OID) ? BTORDER_PROC : HASHPROC); 
}

static AnyArrayTypeInfo*
getAnyArrayTypeInfo(AnyArrayTypeInfo *info, Oid typid)
{
	if (info == NULL)
		info = palloc(sizeof(*info));

	info->typid = typid;
	info->cmpFuncOid = InvalidOid;
	info->hashFuncOid = InvalidOid;

	get_typlenbyvalalign(typid, &info->typlen, &info->typbyval, &info->typalign);

	return info;
}

static SimpleArray*
Array2SimpleArray(AnyArrayTypeInfo	*info, ArrayType *a)
{
	SimpleArray	*s = palloc(sizeof(SimpleArray));

	CHECKARRVALID(a);

	s->info = info;
	if (ARRISVOID(a))
	{
		s->elems = NULL;
		s->nelems = 0;
	}
	else
	{
		deconstruct_array(a, info->typid,
						  info->typlen, info->typbyval, info->typalign,
						  &s->elems, NULL, &s->nelems);
	}

	return s;
}

static ArrayType*
SimpleArray2Array(SimpleArray *s)
{
	return construct_array(s->elems, s->nelems, 
						   s->info->typid, 
						   s->info->typlen, 
						   s->info->typbyval, 
						   s->info->typalign);
}

static int
cmpAscArrayElem(const void *a, const void *b, void *arg)
{
	FmgrInfo	*cmpFunc = (FmgrInfo*)arg;

	return DatumGetInt32(FunctionCall2Coll(cmpFunc, DEFAULT_COLLATION_OID, *(Datum*)a, *(Datum*)b));
}

static int
cmpDescArrayElem(const void *a, const void *b, void *arg)
{
	FmgrInfo	*cmpFunc = (FmgrInfo*)arg;

	return -DatumGetInt32(FunctionCall2Coll(cmpFunc, DEFAULT_COLLATION_OID, *(Datum*)a, *(Datum*)b));
}

static void
sortSimpleArray(SimpleArray *s, int32 direction)
{
	AnyArrayTypeInfo	*info = s->info;
	FmgrInfo			cmpFunc;

	if (!OidIsValid(info->cmpFuncOid))
	{
		info->cmpFuncOid = getAMProc(BTREE_AM_OID, info->typid);

		if (!OidIsValid(info->cmpFuncOid))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("could not find compare function")));
	}

	if (s->nelems > 1)
	{
		fmgr_info(info->cmpFuncOid, &cmpFunc);

		qsort_arg(s->elems, s->nelems, sizeof(Datum), 
				  (direction > 0) ? cmpAscArrayElem : cmpDescArrayElem,
				  &cmpFunc);
	}
}

static void
uniqSimpleArray(SimpleArray *s)
{
	AnyArrayTypeInfo	*info = s->info;
	FmgrInfo			cmpFunc;

	if (!OidIsValid(info->cmpFuncOid))
	{
		info->cmpFuncOid = getAMProc(BTREE_AM_OID, info->typid);

		if (!OidIsValid(info->cmpFuncOid))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("could not find compare function")));
	}

	if (s->nelems > 1)
	{
		Datum	*tmp, *dr;
		int32	cmp, num =  s->nelems;

		fmgr_info(info->cmpFuncOid, &cmpFunc);

		tmp = dr = s->elems;
		while (tmp - s->elems < num)
		{
			cmp = (tmp == dr) ? 0 : cmpAscArrayElem(tmp, dr, &cmpFunc);

			if ( cmp != 0 )
				*(++dr) = *tmp++;
			else
				tmp++;
		}

		s->nelems = dr + 1 - s->elems;
	}
}

PG_FUNCTION_INFO_V1(aa_set);
Datum
aa_set(PG_FUNCTION_ARGS)
{
	Datum 				a = PG_GETARG_DATUM(0);
	Oid 				typid = get_fn_expr_argtype(fcinfo->flinfo, 0);
	AnyArrayTypeInfo	info;
	SimpleArray			s;
	ArrayType			*r;

	getAnyArrayTypeInfo(&info, typid);
	s.nelems = 1;
	s.elems = &a;
	s.info = &info;

	r = SimpleArray2Array(&s);

	PG_RETURN_POINTER(r);
}

PG_FUNCTION_INFO_V1(aa_icount);
Datum
aa_icount(PG_FUNCTION_ARGS)
{
	ArrayType  *a = PG_GETARG_ARRAYTYPE_P(0);
	int32		count;

	CHECKARRVALID(a);
	count = ARRNELEMS(a);
	PG_FREE_IF_COPY(a, 0);

	PG_RETURN_INT32(count);
}

PG_FUNCTION_INFO_V1(aa_sort);
Datum
aa_sort(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	ArrayType			*r;
	text				*dirstr = (fcinfo->nargs == 2) ? PG_GETARG_TEXT_P(1) : NULL;
	int32				direction = 1;
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;

	CHECKARRVALID(a);

	if (dirstr != NULL)
	{
		if (VARSIZE_ANY_EXHDR(dirstr) == 3 &&
			pg_strncasecmp(VARDATA_ANY(dirstr), "asc", 3) == 0)
			direction = 1;
		else if (VARSIZE_ANY_EXHDR(dirstr) == 4 &&
				 pg_strncasecmp(VARDATA_ANY(dirstr), "desc", 4) == 0)
			direction = -1;
		else
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("second parameter must be \"ASC\" or \"DESC\"")));
	}

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	s = Array2SimpleArray(info, a);
	sortSimpleArray(s, direction);
	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	if (dirstr)
		PG_FREE_IF_COPY(dirstr, 1);
	PG_RETURN_POINTER(r);
}

PG_FUNCTION_INFO_V1(aa_sort_asc);
Datum
aa_sort_asc(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	ArrayType			*r;
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	s = Array2SimpleArray(info, a);
	sortSimpleArray(s, 1);
	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	PG_RETURN_POINTER(r);

}

PG_FUNCTION_INFO_V1(aa_sort_desc);
Datum
aa_sort_desc(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	ArrayType			*r;
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	s = Array2SimpleArray(info, a);
	sortSimpleArray(s, -1);
	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	PG_RETURN_POINTER(r);
}

PG_FUNCTION_INFO_V1(aa_uniq);
Datum
aa_uniq(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	ArrayType			*r;
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	s = Array2SimpleArray(info, a);
	uniqSimpleArray(s);
	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	PG_RETURN_POINTER(r);
}

PG_FUNCTION_INFO_V1(aa_idx);
Datum
aa_idx(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	Datum				e = PG_GETARG_DATUM(1);
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;
	FmgrInfo			cmpFunc;
	int					i;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	if (get_fn_expr_argtype(fcinfo->flinfo, 1) != ARR_ELEMTYPE(a))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("array type doesn't match search type")));

	s = Array2SimpleArray(info, a);

	if (!OidIsValid(info->cmpFuncOid))
	{
		info->cmpFuncOid = getAMProc(BTREE_AM_OID, info->typid);

		if (!OidIsValid(info->cmpFuncOid))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("could not find compare function")));
	}

	if (info->typlen < 0)
		e = PointerGetDatum(PG_DETOAST_DATUM(e));

	fmgr_info(info->cmpFuncOid, &cmpFunc);
	for(i=0; i<s->nelems; i++)
	{
		if (cmpAscArrayElem(s->elems + i, &e, &cmpFunc) == 0)
			break;
	}

	PG_FREE_IF_COPY(a, 0);
	if (info->typlen < 0)
		PG_FREE_IF_COPY(DatumGetPointer(e), 1);
	PG_RETURN_INT32( (i < s->nelems) ? (i+1) : 0 );
}

PG_FUNCTION_INFO_V1(aa_subarray);
Datum
aa_subarray(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	ArrayType			*r;
	int32				start = PG_GETARG_INT32(1);
	int32				len = (fcinfo->nargs == 3) ? PG_GETARG_INT32(2) : 0;
	int32				end = 0;
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	s = Array2SimpleArray(info, a);

	if (s->nelems == 0)
	{
		r = SimpleArray2Array(s);

		PG_FREE_IF_COPY(a, 0);
		PG_RETURN_POINTER(r);
	}

	start = (start > 0) ? start - 1 : start;

	if (start < 0)
		start = s->nelems + start;

	if (len < 0)
		end = s->nelems + len;
	else if (len == 0)
		end = s->nelems;
	else
		end = start + len;

	if (end > s->nelems)
		end = s->nelems;

	if (start < 0)
		start = 0;

	if (!(start >= end || end <= 0))
	{
		if (end - start > 0 && start != 0)
			memcpy(s->elems, s->elems + start, (end - start) * sizeof(*s->elems));
		s->nelems = end - start;
	}

	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	PG_RETURN_POINTER(r);
}

PG_FUNCTION_INFO_V1(aa_union_elem);
Datum
aa_union_elem(PG_FUNCTION_ARGS)
{
	ArrayType   		*a = PG_GETARG_ARRAYTYPE_P(0);
	Datum				e = PG_GETARG_DATUM(1);
	SimpleArray			*s;
	AnyArrayTypeInfo	*info;
	ArrayType			*r;
	int 				i;
	FmgrInfo			cmpFunc;

	CHECKARRVALID(a);

	info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;

	if (info == NULL || info->typid != ARR_ELEMTYPE(a))
	{
		fcinfo->flinfo->fn_extra = MemoryContextAlloc(fcinfo->flinfo->fn_mcxt,
													  sizeof(*info));
		info = (AnyArrayTypeInfo*)fcinfo->flinfo->fn_extra;
		info = getAnyArrayTypeInfo(info, ARR_ELEMTYPE(a));
	}

	if (get_fn_expr_argtype(fcinfo->flinfo, 1) != ARR_ELEMTYPE(a))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("array type doesn't match search type")));

	s = Array2SimpleArray(info, a);

	if (!OidIsValid(info->cmpFuncOid))
	{
		info->cmpFuncOid = getAMProc(BTREE_AM_OID, info->typid);

		if (!OidIsValid(info->cmpFuncOid))
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("could not find compare function")));
	}

	if (info->typlen < 0)
		e = PointerGetDatum(PG_DETOAST_DATUM(e));

	if (s->nelems == 0)
		s->elems = palloc(sizeof(*s->elems));
	else
		s->elems = repalloc(s->elems, (s->nelems + 1) * sizeof(*s->elems));

	fmgr_info(info->cmpFuncOid, &cmpFunc);
	for(i=0; i<s->nelems; i++)
	{
		if (cmpAscArrayElem(s->elems + i, &e, &cmpFunc) == 0)
			break;
	}

	if (s->nelems == 0 || i>=s->nelems) {
		s->elems[s->nelems] = e;
		s->nelems++;
	}

	r = SimpleArray2Array(s);

	PG_FREE_IF_COPY(a, 0);
	if (info->typlen < 0)
		PG_FREE_IF_COPY(DatumGetPointer(e), 1);
	PG_RETURN_POINTER(r);
}


