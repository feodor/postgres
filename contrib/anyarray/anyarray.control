# anyarray extension
comment = 'functions, operators, and index support for 1-D arrays'
default_version = '1.0'
module_pathname = '$libdir/anyarray'
relocatable = true
