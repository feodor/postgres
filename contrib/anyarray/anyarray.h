/*-------------------------------------------------------------------------
 * 
 * anyarray.h
 *        anyarray common declarations
 *
 * Copyright (c) 2010-2014, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *        contrib/anyarray/anyarray.c
 *
 *-------------------------------------------------------------------------
 */

#ifndef _ANYARRAY_H_
#define _ANYARRAY_H_
#include "postgres.h"
#include "fmgr.h"
#include "nodes/memnodes.h"
#include "utils/array.h"

typedef struct AnyArrayTypeInfo
{
	Oid				typid;
	int16			typlen;
	bool			typbyval;
	char			typalign;
	MemoryContext	funcCtx;
	Oid				cmpFuncOid;
	bool			cmpFuncInited;
	FmgrInfo		cmpFunc;
	bool			hashFuncInited;
	Oid				hashFuncOid;
	FmgrInfo		hashFunc;
} AnyArrayTypeInfo;

typedef struct SimpleArray
{
	Datum				*elems;
	int32				nelems;
	AnyArrayTypeInfo	*info;
} SimpleArray;

#define NDIM 1
#define ARRISVOID(x)  ((x) == NULL || ARRNELEMS(x) == 0)
#define ARRNELEMS(x)  ArrayGetNItems(ARR_NDIM(x), ARR_DIMS(x))

/* reject arrays we can't handle; but allow a NULL or empty array */
#define CHECKARRVALID(x) \
	do { \
		if (x) { \
			if (ARR_NDIM(x) != NDIM && ARR_NDIM(x) != 0) \
				ereport(ERROR, \
						(errcode(ERRCODE_ARRAY_SUBSCRIPT_ERROR), \
						 errmsg("array must be one-dimensional"))); \
			if (ARR_HASNULL(x) && array_contains_nulls(x)) \
				ereport(ERROR, \
						(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), \
						 errmsg("array must not contain nulls"))); \
		} \
	} while(0)


typedef enum SimilarityType {
	AA_Cosine,
	AA_Overlap
} SimilarityType;

extern SimilarityType SmlType;
extern double SmlLimit;

/*
 * Various support functions
 */
extern AnyArrayTypeInfo* getAnyArrayTypeInfo(MemoryContext ctx, Oid typid);
extern void cmpFuncInit(AnyArrayTypeInfo* info);
extern void hashFuncInit(AnyArrayTypeInfo* info);
extern SimpleArray* Array2SimpleArray(AnyArrayTypeInfo  *info, ArrayType *a);
extern ArrayType* SimpleArray2Array(SimpleArray *s);
extern void sortSimpleArray(SimpleArray *s, int32 direction);
extern void uniqSimpleArray(SimpleArray *s, bool onlyDuplicate);
extern int numOfIntersect(SimpleArray *a, SimpleArray *b);
extern double getSimilarity(SimpleArray *sa, SimpleArray *sb);

#endif

