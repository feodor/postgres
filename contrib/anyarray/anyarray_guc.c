/*-------------------------------------------------------------------------
 * 
 * anyarray_guc.c
 *		GUC management
 *
 * Copyright (c) 2010-2014, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		contrib/anyarray/anyarray_gud.c
 *
 *-------------------------------------------------------------------------
 */

#include "anyarray.h"

#include <utils/guc.h>

SimilarityType	SmlType = AA_Cosine;
double			SmlLimit = 0.6;

static bool AnyArrayInited = false;

static const struct config_enum_entry SmlTypeOptions[] = {
	{"cosine", AA_Cosine, false},
	{"overlap", AA_Overlap, false},
	{NULL, 0, false}
};

static void
initAnyArray()
{
	if (AnyArrayInited)
		return;

	DefineCustomRealVariable(
		"anyarray.similarity_threshold",
		"Lower threshold of array's similarity",
		"Array's with similarity lower than threshold are not similar by % operation",
		&SmlLimit,
		SmlLimit,
		0.0,
		1.0,
		PGC_USERSET,
		0,
		NULL,
		NULL,
		NULL
	);
	DefineCustomEnumVariable(
		"anyarray.similarity_type",
		"Type of similarity formula",
		"Type of similarity formula: cosine(default), overlap",
		(int*)&SmlType,
		SmlType,
		SmlTypeOptions,
		PGC_SUSET,
		0,
		NULL,
		NULL,
		NULL
	);
}

void _PG_init(void);
void
_PG_init(void)
{
	initAnyArray();
}

void _PG_fini(void);
void
_PG_fini(void)
{
	return;
}


