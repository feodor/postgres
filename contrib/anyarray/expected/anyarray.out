SELECT anyset(1234);
 anyset 
--------
 {1234}
(1 row)

SELECT icount('{1234234,0234234}'::int[]);
 icount 
--------
      2
(1 row)

SELECT sort('{1234234,-30,0234234}'::int[]);
         sort         
----------------------
 {-30,234234,1234234}
(1 row)

SELECT sort('{1234234,-30,0234234}'::int[],'asc');
         sort         
----------------------
 {-30,234234,1234234}
(1 row)

SELECT sort('{1234234,-30,0234234}'::int[],'desc');
         sort         
----------------------
 {1234234,234234,-30}
(1 row)

SELECT sort_asc('{1234234,-30,0234234}'::int[]);
       sort_asc       
----------------------
 {-30,234234,1234234}
(1 row)

SELECT sort_desc('{1234234,-30,0234234}'::int[]);
      sort_desc       
----------------------
 {1234234,234234,-30}
(1 row)

SELECT uniq('{1234234,-30,-30,0234234,-30}'::int[]);
           uniq           
--------------------------
 {1234234,-30,234234,-30}
(1 row)

SELECT uniq(sort_asc('{1234234,-30,-30,0234234,-30}'::int[]));
         uniq         
----------------------
 {-30,234234,1234234}
(1 row)

SELECT uniq_d('{1234234,-30,-30,0234234,-30,0234234}'::int[]);
 uniq_d 
--------
 {-30}
(1 row)

SELECT uniq_d(sort_asc('{1234234,-30,-30,0234234,-30,0234234}'::int[]));
    uniq_d    
--------------
 {-30,234234}
(1 row)

SELECT idx('{1234234,-30,-30,0234234,-30}'::int[],-30);
 idx 
-----
   2
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::int[],2,3);
     subarray     
------------------
 {-30,-30,234234}
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::int[],-1,1);
 subarray 
----------
 {-30}
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::int[],0,-1);
         subarray         
--------------------------
 {1234234,-30,-30,234234}
(1 row)

SELECT anyset('1234'::text);
 anyset 
--------
 {1234}
(1 row)

SELECT icount('{1234234,0234234}'::text[]);
 icount 
--------
      2
(1 row)

SELECT sort('{1234234,-30,0234234}'::text[]);
         sort          
-----------------------
 {-30,0234234,1234234}
(1 row)

SELECT sort('{1234234,-30,0234234}'::text[],'asc');
         sort          
-----------------------
 {-30,0234234,1234234}
(1 row)

SELECT sort('{1234234,-30,0234234}'::text[],'desc');
         sort          
-----------------------
 {1234234,0234234,-30}
(1 row)

SELECT sort_asc('{1234234,-30,0234234}'::text[]);
       sort_asc        
-----------------------
 {-30,0234234,1234234}
(1 row)

SELECT sort_desc('{1234234,-30,0234234}'::text[]);
       sort_desc       
-----------------------
 {1234234,0234234,-30}
(1 row)

SELECT uniq('{1234234,-30,-30,0234234,-30}'::text[]);
           uniq            
---------------------------
 {1234234,-30,0234234,-30}
(1 row)

SELECT uniq(sort_asc('{1234234,-30,-30,0234234,-30}'::text[]));
         uniq          
-----------------------
 {-30,0234234,1234234}
(1 row)

SELECT uniq_d('{1234234,-30,-30,0234234,-30,0234234}'::text[]);
 uniq_d 
--------
 {-30}
(1 row)

SELECT uniq_d(sort_asc('{1234234,-30,-30,0234234,-30,0234234}'::text[]));
    uniq_d     
---------------
 {-30,0234234}
(1 row)

SELECT idx('{1234234,-30,-30,0234234,-30}'::text[], '-30');
 idx 
-----
   2
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::text[],2,3);
     subarray      
-------------------
 {-30,-30,0234234}
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::text[],-1,1);
 subarray 
----------
 {-30}
(1 row)

SELECT subarray('{1234234,-30,-30,0234234,-30}'::text[],0,-1);
         subarray          
---------------------------
 {1234234,-30,-30,0234234}
(1 row)

SELECT #'{1234234,0234234}'::int[];
 ?column? 
----------
        2
(1 row)

SELECT '{123,623,445}'::int[] + 1245;
      ?column?      
--------------------
 {123,623,445,1245}
(1 row)

SELECT '{123,623,445}'::int[] + 445;
     ?column?      
-------------------
 {123,623,445,445}
(1 row)

SELECT '{123,623,445}'::int[] + '{1245,87,445}';
         ?column?          
---------------------------
 {123,623,445,1245,87,445}
(1 row)

SELECT '{123,623,445}'::int[] - 623;
 ?column?  
-----------
 {123,445}
(1 row)

SELECT '{123,623,445}'::int[] - '{1623,623}'::int[];
 ?column?  
-----------
 {123,445}
(1 row)

SELECT '{123,623,445}'::int[] | '{1623,623}'::int[];
      ?column?      
--------------------
 {123,445,623,1623}
(1 row)

SELECT '{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,
		21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,
		39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,
		57,58,59,60,61,62,63,64,65}'::int[] | '{1623,623, 32, 60}'::int[];
                                                                                               ?column?                                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,623,1623}
(1 row)

SELECT '{123,623,445}'::int[] & '{1623,623}';
 ?column? 
----------
 {623}
(1 row)

SELECT '{-1,3,1}'::int[] & '{1,2}';
 ?column? 
----------
 {1}
(1 row)

SELECT '{1,3,1}'::int[] & '{1,2}';
 ?column? 
----------
 {1}
(1 row)

SELECT '{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,
		21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,
		39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,
		57,58,59,60,61,62,63,64,65}'::int[] & '{1623,623, 32, 60}'::int[];
 ?column? 
----------
 {32,60}
(1 row)

SELECT #'{1234234,0234234}'::text[];
 ?column? 
----------
        2
(1 row)

SELECT '{123,623,445}'::text[] + '1245'::text;
      ?column?      
--------------------
 {123,623,445,1245}
(1 row)

SELECT '{123,623,445}'::text[] + '445'::text;
     ?column?      
-------------------
 {123,623,445,445}
(1 row)

SELECT '{123,623,445}'::text[] + '{1245,87,445}';
         ?column?          
---------------------------
 {123,623,445,1245,87,445}
(1 row)

SELECT '{123,623,445}'::text[] - '623'::text;
 ?column?  
-----------
 {123,445}
(1 row)

SELECT '{123,623,445}'::text[] - '{1623,623}'::text[];
 ?column?  
-----------
 {123,445}
(1 row)

SELECT '{123,623,445}'::text[] | '{1623,623}'::text[];
      ?column?      
--------------------
 {123,1623,445,623}
(1 row)

SELECT '{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,
		21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,
		39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,
		57,58,59,60,61,62,63,64,65}'::text[] | '{1623,623, 32, 60}'::text[];
                                                                                               ?column?                                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {1,10,11,12,13,14,15,16,1623,17,18,19,2,20,21,22,23,24,25,26,27,28,29,3,30,31,32,33,34,35,36,37,38,39,4,40,41,42,43,44,45,46,47,48,49,5,50,51,52,53,54,55,56,57,58,59,6,60,61,62,623,63,64,65,7,8,9}
(1 row)

SELECT '{123,623,445}'::text[] & '{1623,623}';
 ?column? 
----------
 {623}
(1 row)

SELECT '{-1,3,1}'::text[] & '{1,2}';
 ?column? 
----------
 {1}
(1 row)

SELECT '{1,3,1}'::text[] & '{1,2}';
 ?column? 
----------
 {1}
(1 row)

SELECT '{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,
		21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,
		39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,
		57,58,59,60,61,62,63,64,65}'::text[] & '{1623,623, 32, 60}'::text[];
 ?column? 
----------
 {32,60}
(1 row)

SET anyarray.similarity_type = 'cosine';
SELECT similarity('{3,2}'::text[], '{3,2,1}');
 similarity 
------------
   0.816497
(1 row)

SELECT similarity('{3,2}'::int[], '{3,2,1}');
 similarity 
------------
   0.816497
(1 row)

SELECT similarity('{2,1}'::int[], '{3,2,1}');
 similarity 
------------
   0.816497
(1 row)

SELECT similarity('{}'::int[], '{3,2,1}');
 similarity 
------------
          0
(1 row)

SELECT similarity('{12,10}'::int[], '{3,2,1}');
 similarity 
------------
          0
(1 row)

SELECT similarity('{123}'::int[], '{}');
 similarity 
------------
          0
(1 row)

SELECT similarity('{1,4,6}'::int[], '{1,4,6}');
 similarity 
------------
          1
(1 row)

SELECT similarity('{1,4,6}'::int[], '{6,4,1}');
 similarity 
------------
          1
(1 row)

SELECT similarity('{1,4,6}'::int[], '{5,4,6}'); 
 similarity 
------------
   0.666667
(1 row)

SELECT similarity('{1,4,6}'::int[], '{5,4,6}');
 similarity 
------------
   0.666667
(1 row)

SELECT similarity('{1,2}'::int[], '{2,2,2,2,2,1}');
 similarity 
------------
          1
(1 row)

SELECT similarity('{1,2}'::int[], '{1,2,2,2,2,2}');
 similarity 
------------
          1
(1 row)

SELECT similarity('{}'::int[], '{3}');
 similarity 
------------
          0
(1 row)

SELECT similarity('{3}'::int[], '{3}');
 similarity 
------------
          1
(1 row)

SELECT similarity('{2}'::int[], '{3}');
 similarity 
------------
          0
(1 row)

SET anyarray.similarity_type = 'overlap';
SELECT similarity('{3,2}'::int[], '{3,2,1}');
 similarity 
------------
   0.666667
(1 row)

SET anyarray.similarity_threshold = 0.6;
SELECT '{3,2}'::int[] % '{3,2,1}' AS "true";
 true 
------
 t
(1 row)

SET anyarray.similarity_threshold = 0.7;
SELECT '{3,2}'::int[] % '{3,2,1}' AS "false";
 false 
-------
 f
(1 row)

